src/Pirouette/Symbolic/Eval.hs

```diff
@@ -216,25 +219,34 @@ instance (SymEvalConstr lang) => Alternative (SymEval lang) where
               (runSymEvalRaw env st xs, runSymEvalRaw env st ys)
        in xs' <|> ys'
 
--- | Learn a new constraint and add it as a conjunct to the set of constraints of
---  the current path, then checks whether it leads to an inconsistency and fails
--- if it does.
-learn :: (SymEvalConstr lang) => [C.Constraint lang SymVar] -> SymEval lang ()
-learn cs = do
+pruneOrGetPathProblem :: (SymEvalConstr lang) => [C.Constraint lang SymVar] -> SymEval lang (SolverProblem lang Bool)
+pruneOrGetPathProblem cs = do
   curr <- gets sestConstraint
-  -- Add the conjunct. `C.conjunct` fails when it detects trivial inconsistencies.
   case foldl' (\mc c -> mc >>= C.conjunct c) (Just curr) cs of
     Nothing -> empty
-    Just curr' ->
-      -- As far as `C.conjunct` is concerned, there is no inconsistencies. Let us
-      -- call the solver to actually check that.
-      SymEval $
-        ReaderT $ \env -> do
-          st <- get
-          let st' = st {sestConstraint = curr'}
-          if solvePathProblem (seeSolvers env) (CheckPathProblem st' (seeDefs env))
-            then put st'
-            else empty
+    Just curr' -> do
+      st <- get
+      defs <- asks seeDefs
+      let st' = st {sestConstraint = curr'}
+      put st'
+      return $ CheckPath (CheckPathProblem st' defs)
+
+solveAll :: forall lang res t. Traversable t => SymEval lang (t (SolverProblem lang res)) -> SymEval lang (t res)
+solveAll (SymEval f) = SymEval $
+  ReaderT $ \env -> StateT $ \st -> ListT.Action @IO $ do
+    problems <- ListT.runActions $ runStateT (runReaderT f env) st
+    let -- change the type to get a Traversable instance
+        problemsTraversable = Compose $ Compose . swap <$> problems
+    -- solve the problems asynchronously
+    resultsTraversable <- seeAsyncSolver env @(Compose WeightedList (Compose ((,) (SymEvalSt lang)) t)) problemsTraversable
+    let -- get the correct type back
+        results = swap . getCompose <$> getCompose resultsTraversable
+    return $ ListT.mapWeightedListT (pure . runIdentity) results
+
+learn :: forall lang. (SymEvalConstr lang) => [C.Constraint lang SymVar] -> SymEval lang ()
+learn cs = do
+  result <- fmap runIdentity $ solveAll $ Identity <$> pruneOrGetPathProblem cs
+  unless result empty
```

Comment:
We split `learn` into two sub-functions, `pruneOrGetPathProblem` and `solveAll` (described above). We get a weighted list of results, one for each problem passed to `solveAll`, corresponding to a branch whose constraints we didn't previously find to be inconsistent, and kill off the branches with a negative result (meaning the solver found the constraints to be inconsistent).